### PR TITLE
[JBPM-9555] Process with multiple parallel gateways is not completed in Async mode

### PR DESCRIPTION
**Pull request by elguardian**
Jira: https://issues.redhat.com/browse/JBPM-9555

### CODE REVIEW

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/gateway/ParallelGatewayAsyncTest.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        wih = new ListWorkItemHandler(); 
+        addWorkItemHandler("Human Task", wih);
+        kieSession = createKSession(PARALLEL_GATEWAY_ASYNC);
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        super.tearDown();
+
+        executorService.destroy();
```
Comment (afalhambra) [inline jbpm-test-coverage/src/test/java/org/jbpm/test/functional/gateway/ParallelGatewayAsyncTest.java:67]: would it be worth calling `clearAllRequests` and `clearAllErrors` on the executor before destroying it? wdyt?

**Review by afalhambra — CHANGES_REQUESTED** [review]
Looks good, but just a few changes in the test are requested

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/gateway/ParallelGatewayAsyncTest.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2017 Red Hat, Inc. and/or its affiliates.
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-test-coverage/src/test/java/org/jbpm/test/functional/gateway/ParallelGatewayAsyncTest.java:2]: done

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/gateway/ParallelGatewayAsyncTest.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+    /**
+     * Simple parallel gateway test.
+     */
+    @Test //(timeout = 30000)
+    public void testParallelGatewayAsync() throws Exception {
+        Map<String, Object> inputs = new HashMap<>();
+        inputs.put("useHT", Boolean.TRUE);
+        inputs.put("mode", "1");
+//        kieSession.addEventListener(trackingListener);
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-test-coverage/src/test/java/org/jbpm/test/functional/gateway/ParallelGatewayAsyncTest.java:78]: removed

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/gateway/ParallelGatewayAsyncTest.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        kieSession.addEventListener(new DefaultProcessEventListener() {
+            @Override
+            public void beforeNodeTriggered(ProcessNodeTriggeredEvent event) {
+                System.out.println(event.getNodeInstance().getNode().getNodeUniqueId() + " - " + event.getNodeInstance().getNodeName());
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-test-coverage/src/test/java/org/jbpm/test/functional/gateway/ParallelGatewayAsyncTest.java:82]: removed

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/gateway/ParallelGatewayAsyncTest.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+            }
+
+            
+        });
+        ProcessInstance pi = kieSession.startProcess(PARALLEL_GATEWAY_ASYNC_ID, inputs);
+        Thread.sleep(3000L);
+        wih.getWorkItems().forEach(e -> System.out.println(e));
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-test-coverage/src/test/java/org/jbpm/test/functional/gateway/ParallelGatewayAsyncTest.java:89]: removed

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/gateway/ParallelGatewayAsyncTest.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    /**
+     * Simple parallel gateway test.
+     */
+    @Test //(timeout = 30000)
+    public void testParallelGatewayAsync() throws Exception {
+        Map<String, Object> inputs = new HashMap<>();
+        inputs.put("useHT", Boolean.TRUE);
+        inputs.put("mode", "1");
+//        kieSession.addEventListener(trackingListener);
+        kieSession.addEventListener(new DefaultProcessEventListener() {
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-test-coverage/src/test/java/org/jbpm/test/functional/gateway/ParallelGatewayAsyncTest.java:79]: removed

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/gateway/ParallelGatewayAsyncTest.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    public void tearDown() throws Exception {
+        super.tearDown();
+
+        executorService.destroy();
+    }
+
+    /**
+     * Simple parallel gateway test.
+     */
+    @Test //(timeout = 30000)
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-test-coverage/src/test/java/org/jbpm/test/functional/gateway/ParallelGatewayAsyncTest.java:73]: uncommented

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/gateway/ParallelGatewayAsyncTest.java (hunk 1 above)
Comment (elguardian) [reply jbpm-test-coverage/src/test/java/org/jbpm/test/functional/gateway/ParallelGatewayAsyncTest.java:67]: added. not harm in that.

**Review by afalhambra — APPROVED** [review]
Very good job @elguardian!

**Pull request merged**